nav2_costmap_2d/plugins/costmap_filters/speed_filter.cpp

```diff
@@ -0,0 +1,349 @@
+void SpeedFilter::filterInfoCallback(
+  const nav2_msgs::msg::CostmapFilterInfo::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!mask_sub_) {
+    RCLCPP_INFO(
+      logger_,
+      "SpeedFilter: Received filter info from %s topic.", filter_info_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      logger_,
+      "SpeedFilter: New costmap filter info arrived from %s topic. Updating old filter info.",
+      filter_info_topic_.c_str());
+    // Resetting previous subscriber each time when new costmap filter information arrives
+    mask_sub_.reset();
+  }
+
+  // Set base_/multiplier_ or use speed limit in % of maximum speed
+  base_ = msg->base;
+  multiplier_ = msg->multiplier;
+  if (msg->type == SPEED_FILTER_PERCENT) {
+    // Using speed limit in % of maximum speed
+    RCLCPP_INFO(
+      logger_,
+      "SpeedFilter: Using expressed in a percent from maximum speed"
+      "speed_limit = %f + filter_mask_data * %f",
+      base_, multiplier_);
+  } else if (msg->type == SPEED_FILTER_ABSOLUTE) {
+    // Speed limit in absolute values is not implemented yet
+    RCLCPP_ERROR(logger_, "SpeedFilter: Speed limit in absolute values is not implemented yet");
+    return;
+  } else {
+    RCLCPP_ERROR(logger_, "SpeedFilter: Mode is not supported");
+    return;
+  }
+
+  mask_topic_ = msg->filter_mask_topic;
+
+  // Setting new filter mask subscriber
+  RCLCPP_INFO(
+    logger_,
+    "SpeedFilter: Subscribing to \"%s\" topic for filter mask...",
+    mask_topic_.c_str());
+  mask_sub_ = node->create_subscription<nav_msgs::msg::OccupancyGrid>(
+    mask_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&SpeedFilter::maskCallback, this, std::placeholders::_1));
+}
+
+void SpeedFilter::maskCallback(
+  const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  if (!filter_mask_) {
+    RCLCPP_INFO(
+      logger_,
+      "SpeedFilter: Received filter mask from %s topic.", mask_topic_.c_str());
+  } else {
+    RCLCPP_WARN(
+      logger_,
+      "SpeedFilter: New filter mask arrived from %s topic. Updating old filter mask.",
+      mask_topic_.c_str());
+    filter_mask_.reset();
+  }
+
+  filter_mask_ = msg;
+  mask_frame_ = msg->header.frame_id;
+}
+
+bool SpeedFilter::transformPose(
+  const geometry_msgs::msg::Pose2D & pose,
+  geometry_msgs::msg::Pose2D & mask_pose) const
+{
+  if (mask_frame_ != global_frame_) {
+    // Filter mask and current layer are in different frames:
+    // Prepare transform from current layer frame (global_frame_) to mask frame
+    geometry_msgs::msg::TransformStamped transform;
+    tf2::Transform tf2_transform;
+    tf2_transform.setIdentity();  // initialize by identical transform
+    try {
+      transform = tf_->lookupTransform(
```

Comment:
This if/else block could be simplied if you transformed the pose into a Vector3 and then used TF2 to directly transform the Vector3 from the global frame to the mask frame rather than getting the transformation and applying it manually.

Reply:
Could you please give more guess, what how do you see to optimize?

Currently, it looks like everything that you've listed, already implemented in the code as following:

> if you transformed the pose into a Vector3

```
tf2::Vector3 point(pose.x, pose.y, 0);
```

> and then used TF2 to directly transform the Vector3 from the global frame to the mask frame

```
point = tf2_transform * point;
mask_pose.x = point.x();
mask_pose.y = point.y();
```

Yes, the process of obtaining TF2 looks to be complicated: getting `TransformStamped` by calling `lookupTransform()` from TF2::Buffer and then converting obtained `TransformStamped` into `TF2`. However, when I've searched the rest of navigation2 code, I see the same principal is using in Static Layer and AMCL:
```
nav2_costmap_2d/plugins/static_layer.cpp -> in StaticLayer::updateCosts()
nav2_amcl/src/amcl_node.cpp -> in AmclNode::handleInitialPose()
```
I also found another way in DWB in `nav2_dwb_controller/nav_2d_utils/src/tf_help.cpp` to directly use `tf2::doTransform()` but on `PoseStamped` not on `Vector3` as you've mentioned above. Also, DWB approach seems more heavy for me: there is `PoseStamped` used in transformation and it should have header with timing information filled before transformation. So, I am not sure about it.

Maybe, you know better way how to easy get TF2 from its buffer?

Reply:
Right now, you're getting an abstract transform from TF between 2 frames, then creating a point vector and applying that transformation. Instead, I'm asking you to use TF to transform a specific point vector to the frame within TF to let it handle those details, which it can do. http://wiki.ros.org/tf2/Tutorials/Using%20stamped%20datatypes%20with%20tf2%3A%3AMessageFilter

```
buffer_.transform(stamped_point_in_frame_1, stamped_point_in_frame_target, target_frame_);
```

Reply:
Changed in https://github.com/ros-planning/navigation2/pull/2074/commits/128b9e39f79264a80561ee09e458cc185c6ff5b4